src/providers/TxHistoryProvider.tsx

```diff
@@ -0,0 +1,145 @@
+  // Ensures localStorage is always up to date, so we can persist good data on refresh
+  const _setTransactions = useCallback((txs: TransactionLog[]) => {
+    localStorage.setItem(KEY_TRANSACTIONS, JSON.stringify(txs))
+    setTransactions(txs)
+  }, [])
+
+  // Load initial state
+  useEffect(() => {
+    _setTransactions(
+      JSON.parse(localStorage.getItem(KEY_TRANSACTIONS) || '[]')
+        // Only persist txs that are failed/pending
+        // or were created within history window
+        .filter(
+          (tx: TransactionLog) =>
+            tx.status !== TxStatus.success ||
+            nowSeconds() - TX_HISTORY_TIME_SECS < tx.createdAt,
+        ) as TransactionLog[],
+    )
+  }, [_setTransactions])
+
+  // Setup poller for refreshing transactions
+  useEffect(() => {
+    // Only set new poller if there are pending transactions
+    // Succeeded/failed txs don't need to be refreshed
+    if (!poller && transactions.some(tx => tx.status === TxStatus.pending)) {
+      const threeMinutesAgo = nowSeconds() - 3 * 60
+
+      // If any pending txs were created less than 3 min ago, use short poll time
+      // Otherwise use longer poll time
+      // (Assume no need for quick UX if user has already waited 3 min)
+      const pollInterval = transactions.some(
+        tx => tx.status === TxStatus.pending && threeMinutesAgo < tx.createdAt,
+      )
+        ? SHORT_TERM_POLL_INTERVAL_MILLIS
+        : LONG_TERM_POLL_INTERVAL_MILLIS
+
+      setPoller(
+        setInterval(
+          async () =>
+            _setTransactions(
+              // Refresh all transactions
+              await Promise.all(
+                transactions.map(async txLog => {
+                  // Only do refresh logic for pending txs
+                  if (!txLog.tx.hash || txLog.status !== TxStatus.pending) {
+                    return txLog
+                  }
+
+                  // If no response yet, get response
+                  const response = (txLog.tx as TransactionResponse).wait
+                    ? (txLog.tx as TransactionResponse)
+                    : await readProvider.getTransaction(txLog.tx.hash)
+
+                  let status = TxStatus.pending
+                  try {
+                    await response.wait()
+
+                    // Tx has been mined
+                    status = TxStatus.success
+                  } catch (_) {
+                    // Handle error thrown by ethers provider when a transaction fails
+                    status = TxStatus.failed
+                  }
+
+                  return {
+                    ...txLog,
+                    tx: response,
+                    status,
+                  }
+                }),
+              ),
+            ),
+          pollInterval,
+        ),
+      )
+    }
+
+    // Clean up. Ensures we stop polling once all pending TXs have been refreshed
+    return () => {
+      if (poller) clearInterval(poller)
```

Comment:
we're basically destroying and _conditionally_ recreating the poller every time the state of any transaction changes. i think it's the best way to do this, but i also went kinda mad dancing with react state while testing this so who knows